modules/apps/object/object-dynamic-data-mapping/src/main/resources/META-INF/resources/Autocomplete.js

```diff
@@ -0,0 +1,107 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {useIsMounted} from '@liferay/frontend-js-react-web';
+import {Autocomplete as BaseAutocomplete} from 'frontend-js-components-web';
+import {debounce} from 'frontend-js-web';
+import React, {useState} from 'react';
+
+function getData(apiUrl, query, page, pageSize) {
+	const url = new URL(apiUrl, Liferay.ThemeDisplay.getPortalURL());
+
+	if (query) {
+		url.searchParams.append('search', query);
+	}
+
+	if (page) {
+		url.searchParams.append('page', page);
+	}
+
+	if (pageSize) {
+		url.searchParams.append('pageSize', pageSize);
+	}
+
+	return Liferay.Util.fetch(url, {
+		headers: new Headers({
+			Accept: 'application/json',
+			'Accept-Language': Liferay.ThemeDisplay.getBCP47LanguageId(),
+			'Content-Type': 'application/json',
+		}),
+	}).then((data) => data.json());
+}
+
+function Autocomplete({
+	apiURL,
+	initialLabel,
+	initialValue,
+	inputName,
+	itemsKey,
+	itemsLabel,
+	value,
+}) {
+	const [query, setQuery] = useState(initialLabel || '');
+
+	const [selectedItem, updateSelectedItem] = useState(initialValue || value);
+	const [items, updateItems] = useState(null);
+	const [loading, setLoading] = useState(false);
+	const isMounted = useIsMounted();
+
+	const fetchData = debounce(() => {
+		if (isMounted()) {
+			setLoading(true);
+
+			getData(apiURL, query, 1, 10)
+				.then((jsonResponse) => {
+					updateItems(jsonResponse.items);
```

Comment:
Won't this be unsafe from an `isMounted` point of view? By the time this runs, something else could've happened. I think we have [`useStateSafe`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/hooks/useStateSafe.ts) for this, which will be rendered useless by https://github.com/facebook/react/pull/22114

Reply:
Yeah we probably should use `useStateSafe` for this and I will update it.

From a memory perspective, this isn't a terrible issue because it likely would only happen once and would not be a re-occuring memory leak like an event listener would be. Its sort of "memory splash"™️ 😂 

![](https://media.giphy.com/media/kK7ej3vLS2b8Q/giphy.gif)

Reply:
Actually the more I think about it, and going off what Dan says

>This will trigger the setState warning. However, there is no actual problem here. There's no "memory leak" here: the POST comes back, we try to set state, the component is unmounted (so nothing happens), and then we're done. Nothing keeps holding onto the component indefinitely.


I think we are fine to ignore this case for now and not go overboard with the useSafeState for cases that don't "hold onto the component indefinitely"